text/0000-buildpack-registry-search-api.md

```diff
@@ -0,0 +1,235 @@
+  ```json
+  {
+    "latest": {
+      "description": "The Command Function Buildpack is a Cloud Native Buildpack V3 that provides riff Command Function Invoker to functions",
+      "license": "MIT",
+      "ns":"projectriff",
+      "name":"command-function",
+      "version": "1.4.1",
+      "yanked":false,
+      "addr":"gcr.io/projectriff/command-function@sha256:99f9054abb73635a9b251b61d3627a8ff86508c767f9d691c426d45e8758596f"  
+    },
+    "versions": {
+      "1.4.1": {
+        "link": "https://registry.buildpacks.io/buildpacks/v1/projectriff/command-function/1.4.1"
+      },
+      "1.3.9": {
+        "link": "https://registry.buildpacks.io/buildpacks/v1/projectriff/command-function/1.3.9"
+      }
+    }
+  } 
+
+- **GET /buildpacks/:ns/:name/:version**
+
+  Retrieves metadata for a specific buildpack version (`:version` must be a semver or `latest`).  This response *may* contain more metadata e.g. download metrics.  Since `description` and `license` can change between different versions, it should therefore ONLY be included for each specific buildpack version.
+  ```
+  $ curl https://registry.buildpacks.io/buildpacks/v1/projectriff/command-function/1.4.1
+  ```
+  ```json
+  {
+      "description": "The Command Function Buildpack is a Cloud Native Buildpack V3 that provides riff Command Function Invoker to functions",
+      "license": "MIT",
+      "ns":"projectriff",
+      "name":"command-function",
+      "version": "1.4.1",
+      "yanked":false,
+      "addr":"gcr.io/projectriff/command-function@sha256:99f9054abb73635a9b251b61d3627a8ff86508c767f9d691c426d45e8758596f"
+  }
+  ```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+In the initial implementation, we can take advantage of the existing [registry](https://github.com/buildpacks/registry-index) repository to pull buildpack data from.  To keep ensure up-to-date buildpack data, we'd have a polling loop that performs a `git pull`.  This polling would take place in a separate background process.  
+
+Each time the local repository has been updated via `git pull`, buildpack data will be processed/normalized into a single JSON object, where plain text searches can be used against it.   Fields in this JSON object will be re-indexed as searchable fields, which will be compared against the search text during the retrieval algorithm.
+
+Search fields will include `ns`, and `name`. 
+
+In addition, a server process will expose endpoints and handle incoming request for the GET endpoints mentioned earlier.  For search requests, the text will be extracted from the `query` query parameter and used to search against the normalized buildpacks index.   Results will be added to the server response, as a list of JSON objects.
+
+*Note:  Initially, the Distribution Team maintainers can manage this service, and could even set-up a pager (but only best effort, and during working hours).  In the future, we maybe able to extend this to other verified project contributors.
+
+### Future Work
```

Comment:
We should add the idea of "tying" an API endpoint to a storage location here or under "What related issues do you consider out of scope for this RFC".

The use case in mind is:

> If the user uses a buildpack registry that lives in _some unofficial repo_, how does a tool (such as `pack`) know where to reach the API to search for buildpacks?

Reply:
I don't think we should be worried about `pack` trying to search unofficial registries at this time. There's no guarantee other registries will support the search API, and most unofficial registries (in the near future) will just be Git repos).

Reply:
I have UX concerns about this longterm but can see that we omit this for the time being until it's a little more proven and tested. ...With caution that we don't add constraints/assumptions that make this a lot harder to incorporate in the future.